instrumentation/ruby_kafka/lib/opentelemetry/instrumentation/ruby_kafka/patches/consumer.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+# Copyright 2020 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module RubyKafka
+      module Patches
+        # The Consumer module contains the instrumentation patch for the Consumer class
+        module Consumer
+          def each_message(min_bytes: 1, max_bytes: 10_485_760, max_wait_time: 1, automatically_mark_as_processed: true)
+            super do |message|
+              attributes = {
+                'messaging.system' => 'kafka',
+                'messaging.destination' => message.topic,
+                'messaging.destination_kind' => 'topic',
+                'messaging.kafka.partition' => message.partition
+              }
+
+              attributes['messaging.kafka.message_key'] = message.key if message.key
+              attributes['offset'] = message.offset
+
+              parent_context = OpenTelemetry.propagation.text.extract(message.headers)
+              tracer.in_span('process', with_parent: parent_context, attributes: attributes, kind: :consumer) do
+                yield message
+              end
+            end
+          end
+
+          def each_batch(min_bytes: 1, max_bytes: 10_485_760, max_wait_time: 1, automatically_mark_as_processed: true)
```

Comment:
According the semantic conventions, batch processing should add links: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/messaging.md#batch-processing

Reply:
So this should iterate over `batch.messages` and add a link for each span context extracted? How does this method interact with `each_message` above, if at all?

Reply:
It seems like we should instrument [`fetch_batches`](https://github.com/zendesk/ruby-kafka/blob/0a07531e3a4214a0f8b8b023fa1eff479c535d3f/lib/kafka/consumer.rb#L542-L578) to create a single receive span. I'm unsure how we're supposed to represent `each_batch`, though. Effectively, `KafkaRuby` has a batch receive (with `fetch_batches`) and then either a batch process (`each_batch`) or a single process of each message in a batch (`each_message`). The second case is equivalent to the Batch Receive example in the spec, but the first case is different from the Batch Process example.

If we instrument `fetch_batches`, we would need to propagate the span somehow to the instrumentation of `each_message` and `each_batch`.